src/libAtomVM/unicode.c

```diff
@@ -0,0 +1,48 @@
+/*
+ * This file is part of AtomVM.
+ *
+ * Copyright 2024 Davide Bettio <davide@uninstall.it>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later
+ */
+
+#include <stdbool.h>
+#include <stddef.h>
+
+#include "unicode.h"
+
+size_t unicode_buf_utf8_len(const uint8_t *buf, size_t buf_len)
+{
+    size_t count = 0;
+
+    for (size_t i = 0; i < buf_len; i++) {
+        if ((buf[i] & 0xC0) != 0x80) {
```

Comment:
This warrants a comment.  Not everyone is versed in UTF8 encoding.  Why must the high two bits not be `0x10`?

Reply:
I guess I am a little concerned about this function.  Couldn't the input buffer be an invalid UTF-8 string (for example, suppose it ends in a UTF-8 escape sequence, but has no subsequent UTF-8 encoded characters?).

Wouldn't it be more safe to use `bitstring_utf8_decode` here, to get the number of UTF-8 chars?

Looking at the usage of this function (it's being used in `atom_to_list` and `list_to_atom`), I think it's safe, but if this function is used outside of these functions, it could give erroneous values.

Reply:
1. we agreed on the assumption that internal atom utf8 representation is always valid, so this is not an issue as you mention for `atom_to_list` and `atom_to_binary`
2. `unicode_buf_utf8_len` is meant to be superfast, the result will be valid given a valid input
3. I will overhaul utf8 handling with #1051, I will use a fast and battle tested function for validation and decoding